### ISSUE TITLE
ImageFolder(root) raises error if root contains empty subfolders

### ISSUE THREAD

**Issue by MalteEbner** [pytorch/vision#4925]
### 🐛 Describe the bug

When using `torchvision.datasets.ImageFolder(root)` on a root with a subfolder not containing images, an error is thrown.

#### Example code (works in google colab):

```python
!git clone --depth 1  https://github.com/alexeygrigorev/clothing-dataset-small clothing_dataset_small
!mkdir clothing_dataset_small/empty_subfolder_XYZ
print("############################### Found the following subfolders:")
!ls clothing_dataset_small
print("############################### Trying to create an ImageFolder...")
import torchvision
torchvision.datasets.ImageFolder('clothing_dataset_small')
```

#### Output of example code
```
Cloning into 'clothing_dataset_small'...
remote: Enumerating objects: 3818, done.
remote: Counting objects: 100% (3818/3818), done.
remote: Compressing objects: 100% (3818/3818), done.
remote: Total 3818 (delta 0), reused 3815 (delta 0), pack-reused 0
Receiving objects: 100% (3818/3818), 100.57 MiB | 36.22 MiB/s, done.
############################### Found the following subfolders:
empty_subfolder_XYZ  LICENSE  README.md  test  train  validation
############################### Trying to create an ImageFolder...
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-19-d8d3727bcc2a> in <module>()
      5 print("############################### Trying to create an ImageFolder...")
      6 import torchvision
----> 7 torchvision.datasets.ImageFolder('clothing_dataset_small')

3 frames
/usr/local/lib/python3.7/dist-packages/torchvision/datasets/folder.py in make_dataset(directory, class_to_idx, extensions, is_valid_file)
    100         if extensions is not None:
    101             msg += f"Supported extensions are: {', '.join(extensions)}"
--> 102         raise FileNotFoundError(msg)
    103 
    104     return instances

FileNotFoundError: Found no valid file for the classes .git, empty_subfolder_XYZ. Supported extensions are: .jpg, .jpeg, .png, .ppm, .bmp, .pgm, .tif, .tiff, .webp
```
#### Source of error
https://github.com/pytorch/vision/blob/22ff44fd14139f2a056ad52b9bd109bd958089f3/torchvision/datasets/folder.py#L97-L102
As this check was introduced by @pmeier maybe he can describe the intuitions for this.

### Versions

Collecting environment information...
PyTorch version: 1.10.0+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7.12 (default, Sep 10 2021, 00:21:48)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: False
CUDA runtime version: 11.1.105
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.10.0+cu111
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.11.0
[pip3] torchvision==0.11.1+cu111
[conda] Could not collect

cc @pmeier

**Comment by pmeier** [comment]
Hey @MalteEbner. IIRC, the motivation behind this is to avoid subtle errors. Otherwise you might get a "label gap", because a directory is recognized as category, but has no samples. Take this setup:

```
dataset
├── a
│   └── a.png
├── b
└── c
    └── c.png
```

Removing the check and running

```python
from torchvision.datasets.folder import ImageFolder
import pathlib

dataset = ImageFolder("dataset", loader=lambda path: pathlib.Path(path).name)

for path, category in dataset:
    print(path, category)
```

now prints

```
a.png 0
c.png 2
```

Although you have only two categories, the model handling the data now needs to handle three categories.

**Comment by MalteEbner** [comment]
Hey @pmeier, thanks for your quick answer!

Would it be possible to add an argument `ignore_subdirs_without_images` (by default false) or similar to the constructor that allows to have additional directories without images without throwing the error? I think that having additional non-image folders is quite common, e.g. for `.git`, `venv`, `.ipynb_checkpoints`, labels, or more. See also https://github.com/lightly-ai/lightly/issues/535#issuecomment-954591207

**Comment by pmeier** [comment]
Sure, that would be a good addition. I would prefer something like `ignore_empty` because it is a little shorter. Do you want to send a PR?

**Comment by MalteEbner** [comment]
> Sure, that would be a good addition. I would prefer something like `ignore_empty` because it is a little shorter. Do you want to send a PR?

Yes, I will create a PR with `ignore_empty`.

**Comment by NicolasHug** [comment]
Instead of adding a new parameter, wouldn't it be possible to simple override the `find_classes()` method to ignore empty dirs?

We made this public precisely to handle such custom use-cases and avoid bloating the API with lots of specific parameter.

**Comment by pmeier** [comment]
From an UX perspective, I prefer having a simple switch for general behavior such as this. Overwriting `find_classes` also gives you the freedom to do that, but is less convenient. I would only advocate to do this in one-off use cases.

Plus, in the light of the new datasets, `ImageFolder` will be replaced by a `from_folder` call. There we won't have the ability to overwrite something:

https://github.com/pytorch/vision/blob/b9f4ed93208be9b78615fd8325bf4106e309e540/torchvision/prototype/datasets/_folder.py#L50

**Comment by NicolasHug** [comment]
From a UX perspective, having 2 ways of doing similar things can be confusing. On our side it also adds burden as we need to now maintain the cross-product of all of these features. IMHO this use-case *is* a one-off use case.

> There we won't have the ability to overwrite something:

There probably should then - we can't add a new parameter every time a user asks for a new use-case support

**Comment by MalteEbner** [comment]
> IMHO this use-case _is_ a one-off use case.

I disagree here, as before the latest commit adding the check, this use case was already supported.

> On our side it also adds burden as we need to now maintain the cross-product of all of these features.

I fully agree here, implementing this parameter and passing it adds quite some code. Thus I think it would be easier to change the `FileNotFoundError` to some kind of `EmptySubfoldersWarning`.

**Comment by MalteEbner** [comment]
@pmeier @NicolasHug 
Could you decide on one of the possible solutions to this issue? I don't want to work on anything until this is unclear.
I see the following solutions:
1. Do nothing,  as proposed by @NicolasHug 
2. Implement a parameter `ignore_empty`, as proposed by @pmeier and @MalteEbner 
3. Change the `FileNotFoundError` to a simple warning (`e.g.SubfoldersWithoutImagesWarning`), but don't break, as proposed by @MalteEbner

**Comment by NicolasHug** [comment]
Apart from 1, I would be OK with 

4. Allow sub-directories that do not contain images, without warning, and properly filter-out these folders to avoid the "label gap" issue that @pmeier mentioned https://github.com/pytorch/vision/issues/4925#issuecomment-966994668

This allows more flexibility for the user while maintaining BC, and does not add any new API surface.

**Comment by MalteEbner** [comment]
I like the solution 4. @NicolasHug proposed, what do you think about it @pmeier ?

**Comment by pmeier** [comment]
I'm ok with that.

**Comment by pmeier** [comment]
On a second thought: I think we can only get away with silently filling the label gaps, if `class_to_idx` is generated from the directory structure and not passed explicitly. Otherwise we might introduce subtle bugs, because the predefined categories no longer match the ones used in the dataset.

**Comment by MalteEbner** [comment]
After giving it a little more though it seems to me that there are 3 kinds of subdirectories:
A) empty
B) with files, but without valid files (e.g. a `.git` folder)
C) with valid files (e.g. with `img1.jpg`, `img2.jpg` ...)

Solution 4 proposes to
- raise the `FileNotFoundError` if any subdirectory of type a) exists.
- ignore all subdirectories of type B
- create the dataset with the `class_to_idx` only from subdirectories of type C

As @pmeier pointed out, if `class_to_idx` is passed explicitly, it should be checked if all `class_to_idx.keys()` match a valid subdirectory of type C.

This could be implemented as follows (partly pseudocode, should convey the main idea):
```python
empty_directores = []
non_empty_directories = []
valid_directories = []
for entry in os.scandir(directory):
  if entry.is_dir():
   class_name = entry.name
    subdirectory = os.path.join(directory, class_name)
    directory_type = "empty" # assume the subdirectory is empty, type A
    for root, _, fnames in sorted(os.walk(, followlinks=True)):
        for fname in sorted(fnames):
            directory_type = "non_empty" # the subdirectory has at least one file, type B
            path = os.path.join(root, fname)
            if is_valid_file(path):
                item = path, class_index
                instances.append(item)
                directory_type = "with_valid_files" # the subdirectory has valid files, type C
   if directory_type = "empty":
     empty_directories.append(class_name)
  elif directory_type = "non_empty":
     non_empty_directories.append(class_name)
   else:
     valid_directories.append(class_name)
     
if len(empty_directories) >  0:
  # raise error and name those directories, just as before

if class_to_idx is None:
  class_to_idx = {class: id for id, class in enumerate(valid_directories)}
else:
   if set(class_to_idx.keys()) != set(valid_directories):
       # raise ValueError, distinguish perhaps between the type of non-match
```

**Comment by pmeier** [comment]
I understand it fits your use case, but I currently see no good reason to separate type A) from type B). For `make_dataset` the only thing that should matter if there are any valid files in a directory. If that is not the case, IMO it should be irrelevant if the offending folder was empty or not.

**Comment by NicolasHug** [comment]
Yes, in proposed solution 4. empty folders are part of the "sub-directories that do not contain images" terminology.

Also, maybe I should have made this explicit: if implementing 4 requires more than trivial and short changes, then it's not worth implementing either. Again, users can always use the `find_classes` method.

**Comment by pmeier** [comment]
@NicolasHug Overwriting `find_classes` will only work automatically if you are dealing with actually empty directories. If the folder is not empty but has no valid files, one needs to overwrite and rewrite `make_dataset`, since there is no hook to get around the error. 

Of course if the names of the offending directories are known, overwriting `find_classes` and ignoring them "manually" is sufficient.

**Comment by NicolasHug** [comment]
Can't the user filter out the empty dirs *and* the dirs without images in the `find_classes` function?

**Comment by pmeier** [comment]
Sure, but in order to filter all directories that do not contain valid files, you need to traverse them and look at each file. And this 95% of the work that `make_dataset` does.

**Comment by datumbox** [comment]
Though I understand that there are cases where a dataset might contain empty and non-image folders (Python Environments, Version control, OS specific meta-files/folders, Python notebook checkpoints etc), having these folders in there violates the basic assumption of the API which is that we can scan the existing folders to fetch the labels. Yes it's possible to create a complex solution such as the ones described here to handle corner-cases and ensure we don't introduce a bug. Nevertheless it might also be worth considering that the end user might just have to clean up the dataset directory prior loading it OR subclassing our class to apply custom filtering logic.

**Comment by MalteEbner** [comment]
You have convinced me that I should think of an `ImageFolder` differently: It should not be any folder containing some subdirectories with images, but rather a curated folder with exactly one subdirectory for each class and at least one valid file for each class. Thanks for making this more clear to me.

**Comment by datumbox** [comment]
@MalteEbner Thanks for submitting such a detailed proposal and for being keeping an open-mind throughout the discussion. :)

**Comment by giangnguyen2412** [comment]
Hey @NicolasHug @datumbox @pmeier @MalteEbner @smartkiwi 
What is the solution for this now? I am having the same problem.

**Comment by NicolasHug** [comment]
@giangnguyen2412 IIRC, we went with the status quo: you can define a custom the class discovery logic by overriding the  `find_classes()` method of the dataset class: https://pytorch.org/vision/stable/datasets.html#torchvision.datasets.DatasetFolder.find_classes

**Comment by ohaijen** [comment]
Hello, I am wondering if you would consider reopening this issue, but for a different use case.

Basically, if I understand the discussion correctly, there was a request to allow subdirectories of a folder containing images that are empty, because they may contain something else (e.g. some metadata). In the end it was decided that this is not helpful, because a directory of image data should be carefully curated and not have those kinds of folders.

However, I have a different use case. I am working with the iWildcam dataset (https://wilds.stanford.edu/),  but repackaged in the ImageFolder format. This dataset has the peculiarity that not all classes are present in all splits. So, if I load the test data with ImageFolder and do the same with validation data, the class folder-to-class-id mappings won't line up.

A simple solution here would be to simply have empty class folders for classes not present in some of the splits. But this fails because of the exact safeguards discussed in this thread. It seems to me that an `--allow-empty` argument would be a (generally) helpful and elegant solution for use cases like mine.

**Comment by NicolasHug** [comment]
Thanks for your feedback @ohaijen @AdeelH 

Just to recap this thread for myself, because there's been a ton of different things going on.

```
dataset
├── a
│   └── a.png
├── b
└── c
│   └── c.png
└── config
    └── meta.json
```

Usecase:

- A: Do not consider `b` and `config` as classes and raise an error to avoid the "label gap" issue. This is currently supported as the default behaviour.
- B: Do not consider `b` and `config` as classes, and don't raise an error: just ignore those folders. This is currently supported by allowing a custom `find_classes()` which would filter out `b` and `config`.
- C: Consider `b` to be a class (an empty one). This is what @ohaijen  and @AdeelH need, and this currently not supported at all. There's no way to support that with the current API, this will require a new parameter 
- C': Consider `b` to be an empty class and ignore `config`: this is supported if we supported C, by letting the user override `find_classes()` to filter out `config`.


TL;DR there are legitimate use-cases for allowing empty classes, and for that it should be enough to allow to bypass this error:

https://github.com/pytorch/vision/blob/c8c3839cb9b0fc52885cafa431460f686c46f5ae/torchvision/datasets/folder.py#L98-L99

Not sure there's a cleaner way to do that than to add a new `allow_empty` parameter? @pmeier ?

**Comment by ohaijen** [comment]
Hi @NicolasHug , thanks for addressing this issue! I agree with your recap and proposed solution (For the moment, I have this error commended out, and everything is working fine.)

**Comment by AdeelH** [comment]
>Note that we still need a way to ignore config: for that, find_classes() can be overridden

I like @datumbox's and @MalteEbner's  point above that having irrelevant folders like `config/` should not be allowed. I would echo @MalteEbner's summary with one small change:

>You have convinced me that I should think of an ImageFolder differently: It should not be any folder containing some subdirectories with images, but rather a curated folder with exactly one subdirectory for each class ~and at least one valid file for each class~.

Meaning we simply assume that each folder represents a class and perhaps log a msg/warning noting the number of classes with no samples. No `allow_empty` param needed.

**Comment by NicolasHug** [comment]
Thank you @AdeelH . You're right that if we assumed that every single folder has to be a class folder, then we could just get rid of the `raise` and avoid introducing a new parameter.

One thing that is fairly clear from this thread however, is that there are as many different use-cases as there are users (I'm exaggerating a bit). In other words, if we were to make that assumption, we'd be leaving some users on the side and they'd come and open an issue for their use-cases to be supported.

That's why I prefer going for the option to introduce a new parameter here: it may not provide the absolute best UX across all identified use-cases, but it does support them all in a not too terrible way. And at this point, it seems like the best trade-off to make.

This thread is quite long and to try to make things a bit clearer I'll close it and I've opened https://github.com/pytorch/vision/issues/8297 to keep track of the progress on that task. We're planning to support that for the next 0.18 release. Thanks all for your input.
